### ISSUE TITLE
Refactor harden such that it follows prototype chains

### ISSUE THREAD

**Issue by dckc** [Agoric/SES-shim#104]
What's the purpose / justification of this test?

https://github.com/Agoric/harden/blob/e29849c18e0cfbbd41db1e8765954d4bc51d5e4a/test/test.js#L14

This is the only test that fails when I implement `harden` using [deep freezing](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/xs/preload.md#deep-freezing) from the moddable SDK:

```
export default function harden(root) {
  return Object.freeze(root, true);
}
```
https://github.com/dckc/harden/blob/deepFreeze-xs/src/deepFreeze.js

cc @erights @warner @phoddie

**Comment by dckc** [comment]
@erights writes in https://github.com/Moddable-OpenSource/moddable/issues/399#issuecomment-671625831
> We should **NOT** use that hacky replacement for `harden`.

Well, ok...

> We should use the `harden` from the SES-shim monorepo.

Really? This implementation seems to have a `WeakSet` of roughly all objects used in a program. That's not just a legacy work-around?

> If for some reason it does not work out of the box, we need to fix it. Its semantics are different that the two argument `Object.freeze` in important way.

I guess I need to understand better what that way is. To re-iterate the above from December, "This test seems to be testing an implementation detail, not a feature."

**Comment by dckc** [comment]
I got to talk with @warner about this last night and I see the motivation for the test. More later...

**Comment by erights** [comment]
This is a good time to re-raise this. Now that we got rid of the fringe, since `harden` is now post-lockdown only, we should reconsider whether we should go back to having it be transitive up the prototype chain as it used to be. It would certainly be easier to explain.

**Comment by kriskowal** [comment]
We intend to refactor `harden` such that this issue would go away. We would transitively freeze up the prototype chain.

**Comment by kriskowal** [comment]
(This relaxation is possible because we will continue to have `harden` throw if used before `lockdown`).

**Comment by kriskowal** [comment]
This work will imply necessary changes for inter-realm membranes, which will need to block harden from transitively freezing the other side’s primordials.

**Comment by dckc** [comment]
while tryinng to get XS snapshots (https://github.com/Agoric/agoric-sdk/issues/511) and vat workers (https://github.com/Agoric/agoric-sdk/issues/1845) to work on XS, I ran into:

next challenge:
```
throw new Error(
          `Cannot pass non-frozen objects like ${val}. Use harden()`,
        );
```

some notes from our discussion last Thu (cc @phoddie)

@kriskowal , Can we do makeHardener without the rest of the "patch all the legacy gorp" SES stuff?

kriskowal: You will have to. You should be able to use `@agoric/make-hardener` directly, so long as the fringe is already frozen when you use it. But, cc @erights who would know better.

dckc @warner where is the node part of build-xsnap getting harden from?
ah... build-node.js starts with import '@agoric/install-ses';
how do the new bits get from the SES-shim monorepo to the agoric-sdk monorepo? via npm? is the released make-hardener close enough?

@kriskowal : npm/yarn, yes.
There is a released `@agoric/make-hardener` that ses uses to make harden and put it in global scope.
For the current harden, you'll need to harden a layer at a time down the inheritance hierarchy. I plan to simplify harden so that this will stop being an issue.
Given that you start at Object and work your way down the inheritance hierarchy, it should work. But it is an untried experiment. Please let us know how it goes.

dckc not wonderfully. Segmentation fault. I didn't do any fringe setup, though
oh... maybe that was irrelevant. I think I skipped a step and loaded a bogus snapshot
now we're at `exception: ?.complain: no function!` which I think we earlier decoded to mean "you forgot to provide `testLogger` in your vat endowments"

it's a call to console.error. where are we building console? boy there's a lot to juggle here...
now this is more to be expected:
```
TypeError: prototype function  (){[native code]} of unknown is not already in the fringeSet
```

ok! I got the build.js and deliver-0.js steps to do the snapshot stuff.
Then deliver-1.js throws and then segfaults. hm.

ah... ok... got further:

```
---worker: doProcess: vat.promise[p+5] fulfilledToPresence failed: Error: unknown promiseID 'p+5'
```

I'm using the verboten `Object.freeze(x, true)` though
ideally, XS would have a hardened bit along with the frozen bit in its objects, right?
how does it do Object.frozen? let's see...
it checks XS_DONT_SET_FLAG and XS_DONT_DELETE_FLAG on each property (among other things). So it's not one bit on an object.

@erights : Object.freeze(x, true) has no chance of working. You need to get past the fringeset error so you can do a harden. Try first doing a harden([Object, Function, Array]), to get the super-most layer first. Then hardening of most other things should work because they inherit only from things hardened by the above expression.

@warner : also, if we replaced harden with a dummy version, I don't think anything in swingset would notice the difference

@mfig : Liveslots' marshal layer would notice.

**Comment by erights** [comment]
Fixed by #541
